x-pack/plugin/index-lifecycle/src/main/java/org/elasticsearch/xpack/indexlifecycle/MaintenanceModeUpdateTask.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.indexlifecycle;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.logging.ESLoggerFactory;
+import org.elasticsearch.xpack.core.indexlifecycle.IndexLifecycleMetadata;
+import org.elasticsearch.xpack.core.indexlifecycle.OperationMode;
+
+public class MaintenanceModeUpdateTask extends ClusterStateUpdateTask {
+    private static final Logger logger = ESLoggerFactory.getLogger(MaintenanceModeUpdateTask.class);
+    private final OperationMode mode;
+
+    public MaintenanceModeUpdateTask(OperationMode mode) {
+        this.mode = mode;
+    }
+
+    OperationMode getOperationMode() {
+        return mode;
+    }
+
+    @Override
+    public ClusterState execute(ClusterState currentState) {
+        IndexLifecycleMetadata currentMetadata = currentState.metaData().custom(IndexLifecycleMetadata.TYPE);
+
+        boolean maintenanceModeToChange = currentMetadata.getMaintenanceMode().equals(mode) == false;
+        boolean maintenanceModeRequested = OperationMode.MAINTENANCE_REQUESTED.equals(mode);
+        boolean inMaintenanceMode = OperationMode.MAINTENANCE.equals(currentMetadata.getMaintenanceMode());
+        if ((inMaintenanceMode && maintenanceModeRequested) || maintenanceModeToChange == false) {
+            return currentState;
+        }
```

Comment:
I think we also need to check that if `mode == OperationMode.MAINTENANCE` then `currentMetadata.getMaintenanceMode() == OperationMode.MAINTENANCE_REQUESTED` is also true so we don't end up with a race condition where the user requests `NORMAL` mode while we are iterating through the indices in `IndexLifecycleService` and when it finishes iterating it suddenly goes into `MAINTENANCE` mode ignoring the fact that the user cancelled the request

Reply:
in other words, you can only enter maintenance mode if you were previously in maintenance-requested mode

Reply:
yes

Reply:
and in other words, the state transition for operation modes looks like this

```
<current_mode> → [<next_mode>, ...]

NORMAL → [NORMAL, MAINTENANCE_REQUESTED]
MAINTENANCE_REQUESTED → [NORMAL, MAINTENANCE]
MAINTENANCE → [NORMAL]
```